README.md
# About
A python project to help find the optimal MTU values that maximizes the upload or download speeds between a peer and server. It also helps find bandwidth dead zones caused due to a poor choice of MTUs.

I built the project to help myself find the right MTU values for my WG server and peer. I inadvertently found that the default MTU values for the server and peer in my case put my WG connection in a bandwidth dead zone. [Related reddit post](https://www.reddit.com/r/WireGuard/comments/plm8y7/finding_the_optimal_mtu_for_wg_server_and_wg_peer/).

You can have a look at the real-world heatmaps which are posted by users in the issue [Post your MTU heatmaps here!](https://github.com/nitred/nr-wg-mtu-finder/issues/4) If you happen to successfully use `nr-wg-mtu-finder` and are able to generate a heatmap, please post your heatmap in the issue if possible. 


***Lastly, please read the following documentation carefully!***

* Read the *WARNING* section!
* This project offers no warranties, therefore do not use it in production. Ideally trying using two VMs that are similar to your production setup.
* The project was developed and tested against a WG peer and WG server running Ubuntu 20.04.
    * Additionally the project only supports `python 3.8` and `python 3.9`. While it may work on `python 3.7` and `python 3.10`, you may have to resolve some issues with building depedencies, yourself.  


#### Project Version
```
0.2.1
```


# Example Bandwidth Plot
* Light green and white areas indicate good to optimal MTU values.
* Green areas indicate bad MTU values.
* Dark green areas are dead zones.

![Bandwidth Plot](./examples/example.png)

# Warning
***WARNING: This project contains scripts that run shell commands using root access. DO NOT USE IN PRODUCTION.***

***WARNING: This project tears down and spins up the Wireguard interface in the order of a thousand times. DO NOT USE IN PRODUCTION.***

That being said, if you're an experienced python developer, please go through the code to verify that it meets your security standards.


# Installation

The project has been built and tested on  


Install the following on both the WG server and WG peer
* Install `ping`
    ```bash
    # Ubuntu
    sudo apt install iputils-ping
    ```
* Install `iperf3`
    ```bash
    # Source: https://iperf.fr/iperf-download.php
    ```
* Install `sed`
    ```bash
    # Ubuntu
    sudo apt install sed
    ```
* Install `wg-quick`
    ```bash
    # Should come installed when you install Wireguard
    ```
* Install the project
    ```bash
    # Use your environment manager of choice like virtualenv or conda or poetry to pre-create an environment
    pip install nr-wg-mtu-finder==0.2.1 --upgrade
    ```

# Usage

### Prerequisites
1. Follow the installation instructions above for both WG server and WG peer
1. The project assumes that you already have a working WG installation on both the WG peer and WG server.
1. The project assumes that you already have a WG interface like `wg0`.
1. The project assumes that you already have a WG conf file like `/etc/wireguard/wg0.conf`. ***Take a backup of these files***.
1. Before running the following scripts, the WG interface is expected to be active/online such that the peer is able to ping the server. Use `wg-quick up INTERFACE` on both the WG server and WG peer to activate the connection.
1. Start the WG server script before the WG peer script

### On the WG Server
1. Let your firewall accept connections on port 5201 from IPs within your WG interface. This port is used by the iperf3 server.
   ```text
   # Replace 10.2.0.0/24 with your interface's IP range
   ufw allow proto tcp from 10.2.0.0/24 to any port 5201
   ```
1. Let your firewall accept connections on port 5000 from IPs within your WG interface. This port is used by the flask server.
   ```text
   # Replace 10.2.0.0/24 with your interface's IP range
   ufw allow proto tcp from 10.2.0.0/24 to any port 5000
   ```
1. Add the MTU setting to the WG conf file i.e. `/etc/wireguard/wg0.conf`. Choose any random MTU, it will be replaced by the script anyway:
    ```text
    [Interface]
    ...
    MTU = 1420  # <----- ADD THIS LINE IF NOT ALREADY EXISTS

    [Peer]
    ...
    ```
1. Start the server script with the following command.
    ```bash
    # Example: The script cycles server MTUs from 1280 to 1290 in steps of 2
    nr-wg-mtu-finder --mode server --mtu-min 1280 --mtu-max 1290 --mtu-step 2 --server-ip 10.2.0.1
    ```

### On the WG Peer
1. Add the MTU setting to the WG conf file i.e. `/etc/wireguard/wg0.conf`. Choose any random MTU, it will be replaced by the script anyway:
    ```text
    [Interface]
    ...
    MTU = 1420  # <----- ADD THIS LINE IF NOT ALREADY EXISTS

    [Peer]
    ...
    ```
1. Start the server script with the following command.
    ```bash
    # Example: The script cycles peer MTUs from 1280 to 1290 in steps of 2
    nr-wg-mtu-finder --mode peer --mtu-min 1280 --mtu-max 1290 --mtu-step 2 --server-ip 10.2.0.1
    ```

# How it works?

* Two python scripts need to be running simultaneously, one of the WG server and one on the WG peer. Let's call them *server script* and *peer script*.
* The both scripts use `subprocess.Popen` to run shell commands. The following commands are used and expected to be pre-installed if not already available:
    * `ping`
    * `iperf3`
    * `wg-quick`
    * `sed`
* The server script also runs a `flask` server and the peer script uses `requests` to communicate with the flask server.


### How does the server script work?
1. The flow for the server script is defined in the method `MTUFinder.run_server_mode()`.
1. First, a flask server called a `sync_server` is run is the background on a separate process.
    * The `sync_server's` listens for requests and commands from the peer script so that they can synchronize with each other.
    * The peer script waits for the `sync_server` to be available before running any upload or download tests.
    * The peer scripts get the status and MTU of the server script from the `sync_server`.
    * The peer script tells the `sync_server` that it is done with its cycling through all of its MTUs and is ready for the server script to change its MTU so that it can start a fresh cycle.
    * The `sync_server` informs the peer script that the server script is finished with cycling through all MTUs and that it is going to shut itself down. The peer script uses this signal to shut itself down as well.
1. When the server script receives an `INTIALIZE` signal, it runs the following shell commands
    * First, terminate an `iperf3` server process if it is already running.
    * Spin down the WG interface
        ```
        wg-quick down wg0
        ```
    * Replace the MTU in the WG conf file with the next MTU in the list
        ```
        # 1421 is the new MTU
        sed -i s/MTU.*/MTU = 1421/ /etc/wireguard/wg0.conf
        ```
    * Spin up the WG interface
        ```
        wg-quick up wg0
        ```
    * Run iperf3 in server mode
        ```
        iperf3 -s
        ```
1. If the server has finishing cycling through all of its MTUs and then receives a request from peer script that it is ready for a new cycle, then the server sends a `SHUTDOWN` signal to the peer script via the `sync_server`.


### How does the peer script work?
* On start, the peer script checks if the `sync_server` is reachable. Once it is reachable, it sends a `peer/ready` request to the server script.
* The peer script then waits for the `iperf3` server to start on the server side. Once it recognizes that the iperf3 server has started, and then the peer script starts cycling through each of its MTUs.
    * For each MTU, the peer script runs an upload and download test using the following command
        ```
        # Upload test
        iperf3 -c 10.2.0.1 -J -t 5 -i 5
        # Download test
        iperf3 -c 10.2.0.1 -J -t 5 -i 5 -R
        ```
    * After each download and upload test, the peer script parses the output and stores the bandwidth results in a bandwidth log file.
* Once the peer script is finished cycling through all of its MTU, it sends another `peer/ready` request to the server script and restarts the whole process again with the next server MTU.
* If the server script is finished cycling through all of its MTUs, then it sends a `SHUTDOWN` signal to the peer script as a reply to the `peer/ready` request. The server shuts down after a short delay as does the peer script.
* Finally, the user can check the bandwidth log file to see the results.


### How is the MTU heatmap generated?
* If you successfully ran the server & peer script as described in the instructions, then a log file (csv file) which contains the MTU data like in this [example.csv](https://github.com/nitred/nr-wg-mtu-finder/blob/master/examples/example.csv) is generated by the ***peer script***.
* The filename for this log file looks like `wg_mtu_finder_peer_20220101T000000.csv` and is generated in the same directory where the ***peer script*** was run.
* Once the ***peer script*** is done or is shutting down, then the plot function is called which reads the contents of log csv file and generates a heatmap graph which is written to a png file like in this [example.png](https://github.com/nitred/nr-wg-mtu-finder/blob/master/examples/example.png).
* The filename for the heatmap png looks like `wg_mtu_finder_peer_20220101T000000.png` and is generated in the same directory where the ***peer script*** was run.

So if you successfully ran the server and peer script, you should find two new files (csv + png) generated in the same directory where you ran the ***peer script*** on the ***WG-peer*** server.

# CLI Options

#### nr-wg-mtu-finder
```
$ nr-wg-mtu-finder --help
usage: nr-wg-mtu-finder [-h] --mode MODE --mtu-min MTU_MIN --mtu-max MTU_MAX --mtu-step
                        MTU_STEP --server-ip SERVER_IP [--server-port SERVER_PORT]
                        [--interface INTERFACE] [--conf-file CONF_FILE]
                        [--peer-skip-errors PEER_SKIP_ERRORS]

nr-wg-mtu-finder - Helps find the optimal Wireguard MTU between a WG Server and a WG Peer.

optional arguments:
  -h, --help            show this help message and exit
  --mode MODE           Mode should be 'server' if you are running this script on the WG
                        Server. Mode should be 'peer' if you are running this script on
                        the WG Peer.
  --mtu-min MTU_MIN     Min MTU. Must be in the range [1280, 1500].
  --mtu-max MTU_MAX     Max MTU. Must be in the range [1280, 1500].
  --mtu-step MTU_STEP   By how much to increment the MTU between loops.
  --server-ip SERVER_IP
                        The IP address of the WG server and flask server.
  --server-port SERVER_PORT
                        The port for the flask server.
  --interface INTERFACE
                        The WG interface name. Default: 'wg0'
  --conf-file CONF_FILE
                        The path to the interface config file. Default:
                        '/etc/wireguard/wg0.conf'
  --peer-skip-errors PEER_SKIP_ERRORS
                        Skip errors when known errors occur in 'peer' mode during the MTU
                        loop. The known errors are logged and the loop continues without
                        crashing. Default: 'True'. Example usage: --peer-skip-errors False


```

#### nr-wg-mtu-finder-heatmap
```
$ nr-wg-mtu-finder-heatmap --help
usage: nr-wg-mtu-finder-heatmap [-h] --log-filepath LOG_FILEPATH --heatmap-filepath
                                HEATMAP_FILEPATH

nr-wg-mtu-finder-heatmap - Generate a heatmap file (png) from a log file (csv) that was
created by the `nr-wg-mtu-finder` script. This is useful in case the original script file
crashed midway.

optional arguments:
  -h, --help            show this help message and exit
  --log-filepath LOG_FILEPATH
                        Absolute path to the log file (csv) that was created by the `nr-wg-
                        mtu-finder` script.
  --heatmap-filepath HEATMAP_FILEPATH
                        Absolute path to the heatmap file (png) which will be created from
                        the log file (csv).
         
```

# Development

### Publish to pypi.org
* Bump version
* `pip install poetry==1.1.15`
* `poetry build`
* `poetry publish --dry-run`

# License
MIT


CHANGELOG.md
# Changelog
This project adheres to [Semantic Versioning](http://semver.org/).


## tag: 0.2.1 / 2022-09-06
- (PR #6) Bugfix: switch axis labels for correct heatmap display
- Contributors
  - @ahasbini


## tag: 0.2.0 / 2022-07-10
- Bugfix: Fixed a dependency version issue with `werkzeug`, #3
    - Using `poetry` for package and dependency management.
    - As of this verison, all dependencies are locked.
- Added `nr-wg-mtu-finder-heatmap` script for generating heatmap from the log file.
    - This script comes in handy if the original `nr-wg-mtu-finder` crashes mid way. Then a heatmap can be generated from the partial log file (csv) too, barring formatting issues of the log file.
- Formatted with black (88 linewidth) and isort (profile=black)
- First release to pypi so it can be installed with `pip install nr-wg-mtu-finder==0.2.0 --upgrade`


## tag: 0.1.0 / 2021-09-14
- First working release

nr_wg_mtu_finder/plot.py
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns


def create_heatmap_from_log(log_filepath, heatmap_filepath):
    f, axes = plt.subplots(nrows=2, ncols=2, figsize=(12, 12))

    df = pd.read_csv(log_filepath)

    ax = axes[0, 0]
    dfx = df.pivot(index="server_mtu", columns="peer_mtu", values="upload_rcv_mbps")
    sns.heatmap(
        dfx.values,
        linewidth=0.5,
        ax=ax,
        cmap="Greens_r",
        xticklabels=list(dfx.columns),
        yticklabels=list(dfx.index),
    )
    ax.tick_params(axis="x", rotation=45)
    ax.tick_params(axis="y", rotation=0)
    ax.set(ylabel="Server MTU", xlabel="Peer MTU")
    ax.set_title("Upload Rcv Bandwidth (Mbps)")
    ax.invert_yaxis()

    ax = axes[0, 1]
    dfx = df.pivot(index="server_mtu", columns="peer_mtu", values="upload_send_mbps")
    sns.heatmap(
        dfx.values,
        linewidth=0.5,
        ax=ax,
        cmap="Greens_r",
        xticklabels=list(dfx.columns),
        yticklabels=list(dfx.index),
    )
    ax.tick_params(axis="x", rotation=45)
    ax.tick_params(axis="y", rotation=0)
    ax.set(ylabel="Server MTU", xlabel="Peer MTU")
    ax.set_title("Upload Send Bandwidth (Mbps)")
    ax.invert_yaxis()

    ax = axes[1, 0]
    dfx = df.pivot(index="server_mtu", columns="peer_mtu", values="download_rcv_mbps")
    sns.heatmap(
        dfx.values,
        linewidth=0.5,
        ax=ax,
        cmap="Greens_r",
        xticklabels=list(dfx.columns),
        yticklabels=list(dfx.index),
    )
    ax.tick_params(axis="x", rotation=45)
    ax.tick_params(axis="y", rotation=0)
    ax.set(ylabel="Server MTU", xlabel="Peer MTU")
    ax.set_title("Download Rcv Bandwidth (Mbps)")
    ax.invert_yaxis()

    ax = axes[1, 1]
    dfx = df.pivot(index="server_mtu", columns="peer_mtu", values="download_send_mbps")
    sns.heatmap(
        dfx.values,
        linewidth=0.5,
        ax=ax,
        cmap="Greens_r",
        xticklabels=list(dfx.columns),
        yticklabels=list(dfx.index),
    )
    ax.tick_params(axis="x", rotation=45)
    ax.tick_params(axis="y", rotation=0)
    ax.set(ylabel="Server MTU", xlabel="Peer MTU")
    ax.set_title("Download Send Bandwidth (Mbps)")
    ax.invert_yaxis()

    f.suptitle("Peer MTU vs Server MTU Bandwidth (Mbps)")
    f.tight_layout()
    f.savefig(heatmap_filepath, dpi=300)

    print(
        f"create_heatmap_from_log: Done generating heatmap from log file. Heatmap "
        f"can be found at '{heatmap_filepath}'"
    )


nr_wg_mtu_finder/sync_server.py
from typing import Optional

from flask import Flask, jsonify, request
from typing_extensions import Literal

status: Literal["NOT_INITIALIZED", "INITIALIZED", "SHUTDOWN"] = "NOT_INITIALIZED"
mtu: Optional[int] = None


def run_sync_server(host, port, to_server_queue, from_server_queue):
    """Run a flask/http server which is used to synchronize with the Peer script.

    1. Peer can request the flask/http server for Server MTU and its status.
    2. Peer can request the flask/http server to shutdown once the Peer is finished.
    """
    app = Flask(__name__)

    def shutdown_server():
        shutdown = request.environ.get("werkzeug.server.shutdown")
        if shutdown is None:
            raise RuntimeError("Not running with the Werkzeug Server")
        shutdown()

    @app.route("/server/status", methods=["GET"])
    def server_status():
        global mtu, status
        print("RECEIVED REQUEST /server/status")

        msg = None if to_server_queue.empty() else to_server_queue.get()

        if msg:
            # Update global state
            mtu, status = msg["server_mtu"], msg["server_status"]

            if status == "SHUTDOWN":
                # App will shutdown after sending one last response
                shutdown_server()
                from_server_queue.put("SHUTDOWN")
            elif status == "INITIALIZED":
                pass
            else:
                raise NotImplementedError()

            return jsonify({"server_mtu": mtu, "server_status": status})
        else:
            # Return current state
            return jsonify({"server_mtu": mtu, "server_status": status})

    @app.route("/peer/ready", methods=["GET"])
    def peer_ready():
        """Peer is done with its cycle and is waiting for next cycle."""
        global mtu, status
        print("RECEIVED REQUEST /peer/ready")
        status = "NOT_INITIALIZED"

        from_server_queue.put("INITIALIZE")
        return jsonify({"server_mtu": mtu, "server_status": status})

    app.run(host=host, port=port)


nr_wg_mtu_finder/main_heatmap.py
import argparse
import signal
import sys
import time

from pydantic import BaseModel, StrictStr

from nr_wg_mtu_finder.plot import create_heatmap_from_log


def signal_handler(sig, frame):
    """Handle ctrl+c interrupt.

    Without this handler, everytime a ctrl+c interrupt is received, the server shutdowns
    and proceeds to the next iteration in the loop rather than exiting the program
    altogether.
    """
    print("************Received CTRL-C. Will exit in 1 second************")
    time.sleep(1)
    sys.exit(0)


signal.signal(signal.SIGINT, signal_handler)


class ArgsModel(BaseModel):
    log_filepath: StrictStr
    heatmap_filepath: StrictStr

    class Config:
        orm_mode = True


def setup_args():
    """Setup args."""
    parser = argparse.ArgumentParser(
        description=(
            "nr-wg-mtu-finder-heatmap - "
            "Generate a heatmap file (png) from a log file (csv) that was created "
            "by the `nr-wg-mtu-finder` script. This is useful in case the original "
            "script file crashed midway."
        )
    )
    parser.add_argument(
        "--log-filepath",
        help=(
            "Absolute path to the log file (csv) that was created by the "
            "`nr-wg-mtu-finder` script."
        ),
        required=True,
    )
    parser.add_argument(
        "--heatmap-filepath",
        help=(
            "Absolute path to the heatmap file (png) which will be created from the "
            "log file (csv)."
        ),
        required=True,
    )
    args = parser.parse_args()
    return args


def run():
    args = setup_args()
    args = ArgsModel.from_orm(args)

    create_heatmap_from_log(
        log_filepath=args.log_filepath, heatmap_filepath=args.heatmap_filepath
    )


nr_wg_mtu_finder/main.py
import argparse
import signal
import sys
import time
from distutils.util import strtobool

from pydantic import BaseModel, StrictInt, StrictStr, root_validator
from typing_extensions import Literal

from .mtu_finder import MTUFinder


def signal_handler(sig, frame):
    """Handle ctrl+c interrupt.

    Without this handler, everytime a ctrl+c interrupt is received, the server shutdowns and
    proceeds to the next iteration in the loop rather than exiting the program altogether.
    """
    print("************Received CTRL-C. Will exit in 1 second************")
    time.sleep(1)
    sys.exit(0)


signal.signal(signal.SIGINT, signal_handler)


class ArgsModel(BaseModel):
    mode: Literal["server", "peer"]
    mtu_min: int
    mtu_max: int
    mtu_step: int

    server_ip: StrictStr
    server_port: int = 5000

    peer_skip_errors: bool = True

    interface: StrictStr = "wg0"
    conf_file: StrictStr = "/etc/wireguard/wg0.conf"

    @root_validator(pre=False)
    def validate(cls, values):
        """Generic validations."""
        mtu_min, mtu_max, mtu_step = (
            values.get("mtu_min", None),
            values.get("mtu_max", None),
            values.get("mtu_step", None),
        )

        if not (1280 <= mtu_min <= 1500):
            raise ValueError(f"mtu_min: {mtu_min} must be in range [1280, 1500].")

        if not (1280 <= mtu_max <= 1500):
            raise ValueError(f"mtu_max: {mtu_max} must be in range [1280, 1500].")

        if not (mtu_min <= mtu_max):
            raise ValueError(
                f"mtu_min: {mtu_min} must be less than or equal to mtu_max: {mtu_max}"
            )

        return values

    class Config:
        orm_mode = True


def setup_args():
    """Setup args."""
    parser = argparse.ArgumentParser(
        description=(
            "nr-wg-mtu-finder - Helps find the optimal Wireguard MTU between "
            "a WG Server and a WG Peer."
        )
    )
    parser.add_argument(
        "--mode",
        help=(
            "Mode should be 'server' if you are running this script on the WG Server. "
            "Mode should be 'peer' if you are running this script on the WG Peer."
        ),
        required=True,
    )
    parser.add_argument(
        "--mtu-min",
        help="Min MTU. Must be in the range [1280, 1500].",
        required=True,
    )
    parser.add_argument(
        "--mtu-max",
        help="Max MTU. Must be in the range [1280, 1500].",
        required=True,
    )
    parser.add_argument(
        "--mtu-step",
        help="By how much to increment the MTU between loops.",
        required=True,
    )
    parser.add_argument(
        "--server-ip",
        help="The IP address of the WG server and flask server.",
        required=True,
    )
    parser.add_argument(
        "--server-port",
        help="The port for the flask server. Default: 5000",
        required=False,
        default=5000,
    )
    parser.add_argument(
        "--interface",
        help="The WG interface name. Default: 'wg0'",
        required=False,
        default="wg0",
    )
    parser.add_argument(
        "--conf-file",
        help="The path to the interface config file. Default: '/etc/wireguard/wg0.conf'",
        required=False,
        default="/etc/wireguard/wg0.conf",
    )
    parser.add_argument(
        "--peer-skip-errors",
        help=(
            "Skip errors when known errors occur in 'peer' mode during the MTU loop. "
            "The known errors are logged and the loop continues without crashing. "
            "Default: 'True'. Example usage: --peer-skip-errors False"
        ),
        required=False,
        default=True,
        type=strtobool,
    )
    args = parser.parse_args()
    return args


def run():
    args = setup_args()
    args = ArgsModel.from_orm(args)

    MTUFinder(**args.dict())
